### ISSUE TITLE
24hr graph takes long to draw after period of high client activity

### ISSUE THREAD

**Issue by frollic** [pi-hole/FTL#912]
### Versions
- Pi-hole:  5.1.2
- AdminLTE:  5.1.1
- FTL:  patched version from https://github.com/pi-hole/FTL/issues/868

### Platform

- OS and version:  Centos 7
- Platform:  Oracle cloud

### Expected behavior

After and during a period of large amount of clients, the 24hr activity graph takes "forever" to display.

### Actual behavior / bug

Since my pi-hole is exposted to internet, I somtimes get unwanted guests and behaviours.
The other day there was a DDOS (I'd assume) going on, and the pihole served about 2000
clients during a period of ~24 hrs. 

The DNS request they were making were a know DDOS destination, and I was banning their 
IPs as soon as the requests came in (query [ANY] ..... ), most clients never managed to get 
more than 3 requests in before they were dropped by the FW. 
Never the less, even 1 request would add tho the grand total of clients during the 24hr window.

During the "attack" the 24 activity graph took forever to draw, but it was eventually displayed 
(took several minutes).

2000 clients * 10 requests / (24h * 60min)  = <14 requests/minute, so it shouldn't be a DNS 
performance issue.

If this is expected behaviour or not, I cannot tell, but it is acceptable, due to the amount traffic/clients.

Problem is, when the attack ended the graph *still* takes forever to draw. And when I say ended, I
don't mean right after, but when the whole 24hr DDOS window had passed, and I was back to 15ish
clients or so.

It is fixable by restarting pihole-FTL, but it doesn't sound like the ideal solution.

I'm not sure how the graph is getting drawn/generated, but by the looks of it, the selection beeing
made isn't healthy.

There are no errors anywhere in the logs.

**Comment by DL6ER** [comment]
An open resolver is a bad thing and you know this. I will, nevertheless, stop here and just believe that you know what you're doing so we can continue to check the issue you were facing.

The total-queries diagram data is returned by an actually [very small routine](https://github.com/pi-hole/FTL/blob/dbd4a696afd35efa5e6c4882c74d9102d9d7d6a2/src/api/api.c#L133-L194), only giving you some already precomputed numbers. Neither the number of clients nor the actual number of queries produce any more work for this subroutine.

So my question is now: Do you mean the total number graph or the per-client activity graph?

**Comment by frollic** [comment]
Yeah, we've been down the public DNS road before, not going there again ;)

It's the Client **activity** over last 24 hours one.

The total queries was snappy, no issues there. 

The circular ones *might*  have been slow too, when I think of it.
They started to draw, but might not have finished until the client
activity was done, not entierly sure though, sorry :/

I'm happy to provide you with the pihole-FTL.db, if it helps.

**Comment by DL6ER** [comment]
> It's the Client **activity** over last 24 hours one.

makes a lot more sense.

Thanks for your offer, but I'm afraid the `pihole-FTL.db` file will not be very helpful here as the issue didn't seem to appear after you reloaded queries from your database on a restart.

For reference, the code producing the data for this graph is [here](https://github.com/pi-hole/FTL/blob/master/src/api/api.c#L1125-L1275). A massive slowdown is not expected, however, it may be caused by us sending the data always instantaneously for every time slot and every client so it was about 240 * 2000 = 480,000 individual numbers (string representation) being sent from FTL to PHP where they had to be converted again into integers, processed and sent to your front end. This is a lot less than optimal and we're aware of this.

A lot of the code involved here still comes from a long time ago (~ five years is my estimate right now). It will not be straightforward to fix this and I'm not really looking forward to do this. Not because I don't want to, but because the successor of the FTL -> PHP -> user chain is already pretty far developed (branch `new/http`) where everything in this regard is redone (FTL is directly delivering the data to the user, no detour via PHP).

**Comment by churchofnoise** [comment]
@DL6ER , is running `pihole checkout ftl new/http` sufficient to use this branch?

**Comment by DL6ER** [comment]
Your Pi-hole will work on this branch, however, you will find that the web interface will not be working very well. This is because we're rewriting the entire API into a more consistent REST-style API while we're at redesigning things. Rewriting the API simplifies the web interface code *a lot*, however, it is a tremendous task to rewrite the entirety of scripts grown over the years. We can do it only step-by-step. The corresponding web branch is [`AdminLTE:new/FTL_is_my_new_home`](https://github.com/pi-hole/AdminLTE/tree/new/FTL_is_my_new_home).

When checking out these two branches, you can see where we are so far. I think everything is already in a good shape. The dashboard page is already working. Unfortunately, a lot needs to be redone and other things came in our way. Meanwhile, the web interface advanced and we're currently trying to see when we can allocate a slot to continue working on this topic.

![Screenshot_2020-10-27 Pi-hole - ubuntu-server](https://user-images.githubusercontent.com/16748619/97324407-16a23680-1872-11eb-85f5-3d630a5630b2.png)

**Comment by frollic** [comment]
Thank you for the reply. 

I understand trying to fix the "error" is beyond sub optimal at this point. 
Would it be possible to make some kind of limit option ? 
If the execution takes to long, or if the amount of "individual numbers" is 
too high, the transfer stops ?

Sure, the graph might not complete, but it's still better than having the 
browser hang.

**Comment by DL6ER** [comment]
Oh, if it is the browser hanging, than the problem is likely even somewhere else. I thought it would be the delay FTL -> PHP -> browser where the browser doesn't even receive any data for a long time (= should also not be busy). This is strange, I'll try to see where this may come from.

**Comment by frollic** [comment]
Well,

it's the way it appears from the browser.

FF popus on the Stop / Wait question every XX seconds, becasue the graph takes too long to draw.
